### DISCUSSION TITLE
Actions conditional expression not being evaluated correctly (step run when should evaluate to false)

### DISCUSSION THREAD

**Discussion by jhollowe** [community/community#35276] [Actions] [answered]
When trying to turn off an GH Actions step if two version numbers are the same, the Actions debug log shows the evaluation reducing down to `1.2 != 1.2` (which is false) and yet the step still executes. According to [the documentation](https://docs.github.com/en/actions/learn-github-actions/expressions#operators), even if they are different types (like one is a string and one is a number), the types will be converted to numbers when in a comparison/equation statement.

Any ideas as to why this step is being executed? 

[Log](https://github.com/jhollowe/docs/actions/runs/3186864999/jobs/5197869708#step:15:40)
[Workflow File](https://github.com/jhollowe/docs/blob/88cb27a5134d23308f2c0eeb3e28681f7caee4ca/.github/workflows/gh-pages_tags.yaml)

**Comment by vitalyk-multinarity** [comment]
I suggest to fix your workflow file example - currently it doesn't include any 'if'.

**Reply by jhollowe** [reply]
updated, thanks! I had linked in the wrong file.

**Comment by vitalyk-multinarity** [comment]
I didn't try your workflow end-to-end, but I agree - it seems like a bug with type conversion.
 I suggest opening an issue.

**Reply by jhollowe** [reply]
Where can I open an issue for this?

**Reply by airtower-luna** [reply]
@vitalyk-multinarity That category is for issues with the issues feature. :wink:

**Reply by vitalyk-multinarity** [reply]
Try https://github.com/orgs/community/discussions/categories/issues

**Comment by airtower-luna** [comment, marked as answer]
Here's your bug:

```yaml
        if: ${{ steps.versions.outputs.mike }} != ${{ steps.versions.outputs.tag }}
```

Everything *not* enclosed in `${{ ... }}` is interpreted as a literal string. After the actions expressions are substituted you get:

```
'1.2 != 1.2'
```

This is a *string*, not a boolean. And a non-empty string evaluates to `true`. The `!=` needs to be part of the expression to work:

```yaml
        if: ${{ steps.versions.outputs.mike != steps.versions.outputs.tag }}
```

**Reply by vitalyk-multinarity** [reply]
Wow, thanks! Good to know, very non-intuitive for me...

**Reply by jhollowe** [reply]
Ah makes sense now. Thank you so much!
